### PR TITLE
Add RLOO (REINFORCE Leave-One-Out) learner for lower-variance policy …

### LINKED ISSUE #953: Feature Request: RLOO (REINFORCE Leave-One-Out) Advantage Estimator
## Is your feature request related to a problem? Please describe.

I'm working on the Tunix hackathon and running into stability issues training on rubric-based rewards (for the "show your work" objective). The current GRPO works well when you have binary correct/wrong signals from a verifier, but my setup uses continuous scores from a reward model evaluating reasoning quality.

The problem is GRPO's std normalization amplifies noise when rewards are subjective. I'm also seeing reward hacking since there's no KL mechanism to keep the policy from drifting too far to exploit the reward model.

## Describe the solution you'd like

Add RLOO (REINFORCE Leave-One-Out) as an alternative advantage estimator. The main difference from GRPO:

```python
# GRPO (current)
A_i = (R_i - mean(R)) / std(R)

# DrGRPO (removes std, already exists)
A_i = R_i - mean(R)

# RLOO (leave-one-out baseline + KL in reward)
A_i = R_i - mean(R_j where j != i) - β * KL_i
```

RLOO uses a leave-one-out baseline instead of batch mean, which is more numerically stable. Critically, it folds KL directly into the reward `R'_i = R_i - β * KL_i` before advantage computation, rather than adding it to the loss function afterward.

[Ahmadian et al. 2024](https://arxiv.org/abs/2402.14740) showed RLOO is more robust to noisy rewards than PPO/GRPO and outperforms DPO on preference tasks. It's also what [PRIME](https://arxiv.org/abs/2502.01456) uses under the hood.

## Implementation approach

Verified that Tunix already has the infrastructure for this (function_registry.py):

1. Add `advantage_estimator='rloo'` option in `GRPOConfig`
2. Add `kl_in_reward: bool = False` parameter to control where KL is applied
3. Register new advantage estimator function with `@function_registry.register_advantage_estimator("rloo")` (following drgrpo_learner.py pattern)
4. Modify reward computation in `_generate_and_compute_advantage()` to optionally fold KL into rewards when `kl_in_reward=True`

This follows the existing pluggable advantage estimator pattern and doesn't require a separate learner class.

## Additional context

This would help hackathon participants working on non-verifiable tasks. Right now GRPO is tuned for math/code with binary rewards - RLOO would extend Tunix to creative/subjective domains.

Reference implementations:
- [verl RLOO](https://github.com/volcengine/verl)
- [swift RLOO docs](https://swift.readthedocs.io/en/latest/Instruction/GRPO/AdvancedResearch/RLOO.html)

## Checklist

- [x] I have searched the existing issues for similar feature requests.
- [x] I have verified the codebase already supports pluggable advantage estimators via function_registry.
- [x] This is not a support question (please use the "bug template" for that).

### PR DESCRIPTION
**Pull request by kbhujbal**
## Summary

Implements the RLOO (REINFORCE Leave One Out) advantage estimator as a first class learner in Tunix, addressing #953.

RLOO replaces GRPO's group mean advantage baseline with a leave one out baseline: for each completion, the baseline is the mean reward of all *other* completions to the same prompt. This yields **unbiased advantage estimates with lower variance** than standard GRPO, without requiring a separate critic network (unlike PPO).

## Changes

- **`tunix/rl/grpo/rloo_learner.py`** — `RLOOConfig` and `RLOOLearner` following the same pattern as `DrGRPOConfig`/`DrGRPOLearner`
- **`tunix/rl/algorithm_config.py`** — Register `"rloo"` and `"drgrpo"` as valid algo variants and advantage estimators
- **`tunix/cli/grpo_main.py`** — Route to `RLOOLearner` when `advantage_estimator=rloo` is set in `grpo_config`
- **`tunix/cli/base_config.yaml`** — Document `advantage_estimator` and `loss_agg_mode` config options
- **`tests/rl/grpo/rloo_learner_test.py`** — Comprehensive tests (config, advantage math, e2e training)
- **`examples/rl/rloo/gsm8k/run_qwen3.sh`** — Example training script for GSM8K with Qwen3

### Usage

Switch from GRPO to RLOO with a single config change:

```bash
grpo_config.advantage_estimator=rloo
```

Or via Python API:

```python
from tunix.rl.grpo.rloo_learner import RLOOConfig, RLOOLearner

config = RLOOConfig(num_generations=8, beta=0.04, epsilon=0.2)
learner = RLOOLearner(rl_cluster=cluster, algo_config=config, reward_fns=reward_fn)
learner.train(dataset)
```

Resolves #953

**Reference**

Ahmadian et al., ["Back to Basics: Revisiting REINFORCE-Style Optimization for Learning from Human Feedback in LLMs"](https://arxiv.org/abs/2402.14740), 2024.

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

File: examples/rl/rloo/gsm8k/run_qwen3.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Copyright 2025 Google LLC
```
Comment (sizhit2) [inline examples/rl/rloo/gsm8k/run_qwen3.sh]: 2026?

File: tests/rl/grpo/rloo_learner_test.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+from tunix.rl.grpo import grpo_learner as grpo_lib
+from tunix.rl.grpo import rloo_learner as rloo_lib
+from tunix.rl.rollout import base_rollout
+from tunix.tests import test_common as tc
+
+os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=2'
+
+Mesh = sharding.Mesh
+
+_DUMMY_DATA = [
```
Comment (sizhit2) [inline tests/rl/grpo/rloo_learner_test.py]: I believe many of the test helper function could be reused to reduce maintenance cost.

**Review by sizhit2 — COMMENTED** [review]
Thanks for the contribution! Leave some comments.

File: tunix/cli/grpo_main.py [hunk 3]
```diff
@@ -290,11 +292,28 @@ def run_grpo_trainer(self):
...
     rl_cluster = self.create_rl_cluster(tokenizer)
-    grpo_trainer = grpo_learner.GrpoLearner(
-        rl_cluster=rl_cluster,
-        reward_fns=self.obtain_reward_fn(),
-        algo_config=GrpoConfig(**self.config["grpo_config"]),
-    )
+    grpo_config = self.config["grpo_config"]
+    advantage_estimator = grpo_config.get("advantage_estimator", "grpo")
+
+    if advantage_estimator == "rloo":
```
Comment (sizhit2) [inline tunix/cli/grpo_main.py]: Let's handle for all the other possible advantage estimator and error out properly.

File: examples/rl/rloo/gsm8k/run_qwen3.sh (hunk 1 above)
Comment (kbhujbal) [reply examples/rl/rloo/gsm8k/run_qwen3.sh]: Fixed. updated all new files to Copyright 2026 Google LLC.

File: tests/rl/grpo/rloo_learner_test.py (hunk 2 above)
Comment (kbhujbal) [reply tests/rl/grpo/rloo_learner_test.py]: Good call. Extracted the shared toy model setup, dummy datasets, and reward functions into a new tunix/tests/grpo_test_base.py (mirroring the lora_params_test_base.py pattern). The RLOO test now imports from there instead of duplicating  ~80 lines.

File: tunix/cli/grpo_main.py (hunk 3 above)
Comment (kbhujbal) [reply tunix/cli/grpo_main.py]: Done. Replaced the if/else with a _STANDARD_GRPO_LEARNERS dispatch table mapping grpo/rloo/drgrpo to their (module, config, learner) triples. Unknown values now raise a ValueError listing the supported estimators. Also wired up DrGRPOLearner since it was already a valid choice.
